{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "hD1K2juXn-uT"
   },
   "source": [
    "# ZHUNT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-17T12:58:56.666169Z",
     "iopub.status.busy": "2025-06-17T12:58:56.665805Z",
     "iopub.status.idle": "2025-06-17T12:58:56.687660Z",
     "shell.execute_reply": "2025-06-17T12:58:56.685986Z",
     "shell.execute_reply.started": "2025-06-17T12:58:56.666137Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "zhunt_impl = r\"\"\"\n",
    "/*\n",
    "Z-HUNT-2 computer program, Sept.19, 1990\n",
    "*/\n",
    "\n",
    "/*\n",
    "serialized i/o to allow to run against large datasets \n",
    "Updated i/o permissions to match posix. campt 1/10/2000\n",
    "*/\n",
    "\n",
    "/*\n",
    "Turbo C compiler Ver. 1.5\n",
    "Compact model, 8086/80286 instruction set, math emulator/8087/80287\n",
    "Speed option, unsigned char\n",
    "Run on IBM PC XT/AT or compatibles\n",
    "*/\n",
    "\n",
    "/*\n",
    "Written by Ping-jung Chou, under the instruction of Pui S. Ho, according to the\n",
    "paper \"A computer aided thermodynamic approach for predicting the formation of\n",
    "Z-DNA in naturally occurring sequences\",\n",
    "The EMBO Journal, Vol.5, No.10, pp2737-2744, 1986\n",
    "With 0.22 kcal/mol/dinuc for mCG (Zacharias et al, Biochemistry, 1988, 2970)\n",
    "*/\n",
    "\n",
    "\n",
    "#include <stdlib.h>\n",
    "#include <stdio.h>\n",
    "#include <math.h>\n",
    "#include <string.h>\n",
    "#include <time.h>\n",
    "\n",
    "/* creates a temporary file and mmaps the sequence into it */\n",
    "#ifdef USE_MMAP\n",
    "#include <sys/types.h>\n",
    "#include <sys/mman.h>\n",
    "#include <sys/types.h>\n",
    "#include <sys/stat.h>\n",
    "#include <fcntl.h>\n",
    "#endif\n",
    "\n",
    "\n",
    "double linear_search( double x1, double x2, double tole, double (*func)() );\n",
    "double delta_linking( double dl );\n",
    "double delta_linking_slope( double dl );\n",
    "void   show_probability( unsigned seqlength, float *dl, float *slope, float *probability, char *sequence, char **antisyn, char *filename, int low, int high);\n",
    "\n",
    "double linear_search( double x1, double x2, double tole, double (*func)() )\n",
    "{\n",
    "  double f, fmid, dx, xmid, x;\n",
    "\n",
    "  f = func( x1 );\n",
    "  fmid = func( x2 );\n",
    "  if( f * fmid >= 0.0 )\n",
    "    return x2;\n",
    "  x = (f < 0.0) ? (dx = x2 - x1, x1) : (dx = x1 - x2, x2);\n",
    "  do\n",
    "    {\n",
    "      dx *= 0.5;\n",
    "      xmid = x + dx;\n",
    "      fmid = func( xmid );\n",
    "      if( fmid <= 0.0 )\n",
    "        x = xmid;\n",
    "    }  while( fabs( dx ) > tole );\n",
    "  return x;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "int    terms;\n",
    "double *bztwist, *logcoef, *exponent;\n",
    "double _k_rt = -0.2521201;              /* -1100/4363 */\n",
    "double sigma = 16.94800353;             /* 10/RT */\n",
    "double deltatwist;\n",
    "double explimit = -600.0;\n",
    "int low, high ;\n",
    "\n",
    "\n",
    "double delta_linking( double dl )\n",
    "{\n",
    "  double sump, sumq, z, expmini;\n",
    "  int    i;\n",
    "\n",
    "  expmini = 0.0;\n",
    "  for( i=0; i<terms; i++ )\n",
    "    {\n",
    "      z = dl - bztwist[i];\n",
    "      exponent[i] = z = logcoef[i] + _k_rt * z * z;\n",
    "      if( z < expmini )\n",
    "        expmini = z;\n",
    "    }\n",
    "  expmini = (expmini < explimit) ? explimit - expmini : 0.0;\n",
    "  sump = sumq = 0.0;\n",
    "  for( i=0; i<terms; i++ )\n",
    "    {\n",
    "      z = exp( exponent[i] + expmini );\n",
    "      sumq += z;\n",
    "      sump += bztwist[i] * z;\n",
    "    }\n",
    "  sumq += exp( _k_rt * dl * dl + sigma + expmini );\n",
    "  return deltatwist - sump / sumq;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "double delta_linking_slope( double dl )\n",
    "{\n",
    "  double sump, sump1, sumq, sumq1, x, y, z, expmini;\n",
    "  int    i;\n",
    "\n",
    "  expmini = 0.0;\n",
    "  for( i=0; i<terms; i++ )\n",
    "    {\n",
    "      z = dl - bztwist[i];\n",
    "      exponent[i] = z = logcoef[i] + _k_rt * z * z;\n",
    "      if( z < expmini )\n",
    "        expmini = z;\n",
    "    }\n",
    "  expmini = (expmini < explimit) ? explimit - expmini : 0.0;\n",
    "  sump = sump1 = sumq = sumq1 = 0.0;\n",
    "  x = 2.0 * _k_rt;\n",
    "  for( i=0; i<terms; i++ )\n",
    "    {\n",
    "      z = dl - bztwist[i];\n",
    "      y = exp( exponent[i] + expmini );\n",
    "      sumq += y;\n",
    "      sump += bztwist[i] * y;\n",
    "      y *= z * x;\n",
    "      sumq1 += y;\n",
    "      sump1 += bztwist[i] * y;\n",
    "    }\n",
    "  y = exp( _k_rt * dl * dl + sigma + expmini );\n",
    "  sumq += y;\n",
    "  sumq1 += x * dl * y;\n",
    "  return (sump1 - sump * sumq1 / sumq) / sumq;\n",
    "}                                       /* slope at delta linking = dl */\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "                                        /* Delta BZ Energy of Dinucleotide */\n",
    "double dbzed[4][17] = {\n",
    "/* AS-AS */\n",
    "{ 4.40, 6.20, 3.40, 5.20, 2.50, 4.40, 1.40, 3.30, 3.30, 5.20, 2.40, 4.20, 1.40, 3.40, 0.66, 2.40 ,4.26},\n",
    "/* SA-SA */\n",
    "{ 4.40, 2.50, 3.30, 1.40, 6.20, 4.40, 5.20, 3.40, 3.40, 1.40, 2.40, 0.66, 5.20, 3.30, 4.20, 2.40 ,4.26},\n",
    "/* AS-SA */\n",
    "{ 6.20, 6.20, 5.20, 5.20, 6.20, 6.20, 5.20, 5.20, 5.20, 5.20, 4.00, 4.00, 5.20, 5.20, 4.00, 4.00 ,4.26 },\n",
    "/* SA-AS */\n",
    "{ 6.20, 6.20, 5.20, 5.20, 6.20, 6.20, 5.20, 5.20, 5.20, 5.20, 4.00, 4.00, 5.20, 5.20, 4.00, 4.00 ,4.26 }\n",
    "                      };\n",
    "\n",
    "double expdbzed[4][17];                 /* exp(-dbzed/rt) */\n",
    "int    *bzindex;                        /* dinucleotides */\n",
    "\n",
    "\n",
    "\n",
    "int      user_regret( void );\n",
    "FILE     *open_file( int mode, char *filename, char *typestr );\n",
    "void     assign_bzenergy_index( int nucleotides, char seq[] );\n",
    "void     best_anti_syn( int dinucleotides, float esum );\n",
    "void     anti_syn_energy(int dinculeotide, int dinucleotides, float esum );\n",
    "unsigned input_sequence( FILE *file, int nucleotides, int showfile );\n",
    "double   assign_probability( double dl );\n",
    "\n",
    "\n",
    "\n",
    "void assign_bzenergy_index( int nucleotides, char seq[] )\n",
    "{\n",
    "  int  i, j, idx;\n",
    "  char c1, c2;\n",
    "\n",
    "  i = j = 0;\n",
    "  do\n",
    "    {\n",
    "      c1 = seq[i++];\n",
    "      c2 = seq[i++];\n",
    "      switch( c1 )\n",
    "        {\n",
    "          case 'a' : switch( c2 )\n",
    "                       {\n",
    "                         case 'n' : idx = 16;  break;\n",
    "\t\t\t case 'a' : idx = 0;  break;\n",
    "                         case 't' : idx = 1;  break;\n",
    "                         case 'g' : idx = 2;  break;\n",
    "                         case 'c' : idx = 3;  break;\n",
    "                       }  break;\n",
    "          case 't' : switch( c2 )\n",
    "                       {\n",
    "                         case 'n' : idx = 16;  break;\n",
    "\t\t\t case 'a' : idx =  4;  break;\n",
    "                         case 't' : idx =  5;  break;\n",
    "                         case 'g' : idx =  6;  break;\n",
    "                         case 'c' : idx =  7;  break;\n",
    "                       }  break;\n",
    "          case 'g' : switch( c2 )\n",
    "                       {\n",
    "                         case 'n' : idx = 16;  break;\n",
    "\t\t\t case 'a' : idx =  8;  break;\n",
    "                         case 't' : idx =  9;  break;\n",
    "                         case 'g' : idx = 10;  break;\n",
    "                         case 'c' : idx = 11;  break;\n",
    "                       }  break;\n",
    "          case 'c' : switch( c2 )\n",
    "                       {\n",
    "                         case 'n' : idx = 16;  break;\n",
    "\t\t\t case 'a' : idx = 12;  break;\n",
    "                         case 't' : idx = 13;  break;\n",
    "                         case 'g' : idx = 14;  break;\n",
    "                         case 'c' : idx = 15;  break;\n",
    "                       }  break;\n",
    "          case 'n' : switch( c2 )\n",
    "                       {\n",
    "                         case 'a' : idx = 16;  break;\n",
    "                         case 't' : idx = 16;  break;\n",
    "                         case 'g' : idx = 16;  break;\n",
    "\t\t\t case 'n' : idx = 16;  break;\n",
    "                         case 'c' : idx = 16;  break;\n",
    "                       }  break;\t\t\t\t\t   \n",
    "\t\t\t\t\t   \n",
    "        }\n",
    "      bzindex[j++] = idx;\n",
    "    }  while( i < nucleotides );\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "double *bzenergy, *best_bzenergy;       /* dinucleotides */\n",
    "float  best_esum;               /* assigned before call to anti_syn_energy() */\n",
    "char   *best_antisyn, *antisyn;         /* nucleotides */\n",
    "\n",
    "\n",
    "\n",
    "void best_anti_syn( int dinucleotides, float esum )\n",
    "{\n",
    "  int i;\n",
    "  double dl, slope;\n",
    "\n",
    "  if( esum < best_esum )\n",
    "    {\n",
    "      best_esum = esum;\n",
    "      for( i=0; i<dinucleotides; i++ )\n",
    "        best_bzenergy[i] = bzenergy[i];\n",
    "      strcpy( best_antisyn, antisyn );\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "void anti_syn_energy( int din, int dinucleotides, float esum )\n",
    "{\n",
    "  int   i, nucleotides;\n",
    "  float e;\n",
    "\n",
    "  nucleotides = 2 * din;\n",
    "\n",
    "  antisyn[nucleotides] = 'A';\n",
    "  antisyn[nucleotides+1] = 'S';\n",
    "  i = (din == 0) ? 0 : ((antisyn[nucleotides-1] == 'S') ? 0 : 3);\n",
    "  e = dbzed[i][bzindex[din]];\n",
    "  esum += e;\n",
    "  bzenergy[din] = expdbzed[i][bzindex[din]];\n",
    "  if( ++din == dinucleotides )\n",
    "    best_anti_syn( dinucleotides, esum );\n",
    "  else\n",
    "    anti_syn_energy( din, dinucleotides, esum );\n",
    "  esum -= e;\n",
    "  din --;\n",
    "\n",
    "  antisyn[nucleotides] = 'S';\n",
    "  antisyn[nucleotides+1] = 'A';\n",
    "  i = (din == 0) ? 1 : ((antisyn[nucleotides-1] == 'A') ? 1 : 2);\n",
    "  esum += dbzed[i][bzindex[din]];\n",
    "  bzenergy[din] = expdbzed[i][bzindex[din]];\n",
    "  if( ++din == dinucleotides )\n",
    "    best_anti_syn( dinucleotides, esum );\n",
    "  else\n",
    "    anti_syn_energy( din, dinucleotides, esum );\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "char *tempstr, *sequence;\n",
    "#ifdef USE_MMAP\n",
    "int sequencefile;\n",
    "#endif\n",
    "\n",
    "int user_regret( void )\n",
    "{\n",
    "  char c;\n",
    "  do\n",
    "    {\n",
    "      gets( tempstr );\n",
    "      c = tempstr[0];\n",
    "    }  while( c == 0 );\n",
    "  return (c == '@') ? 1 : 0;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "FILE *open_file( int mode, char *filename, char *typestr )\n",
    "{\n",
    "  static char *iostr[] = { \"output\", \"input\" };\n",
    "  static char *rwstr[] = { \"w\",     \"r\" };\n",
    "  char  *fullfile;\n",
    "  \n",
    "  FILE *file;\n",
    "  file = NULL;\n",
    "  fullfile = (char *) malloc(sizeof(char) * ( strlen(filename) + strlen(typestr) + 1) );\n",
    "  strcpy(fullfile, filename);\n",
    "  if(strlen(typestr) != 0)\n",
    "\t{\n",
    "        strcat(fullfile, \".\");\n",
    "\tstrcat(fullfile, typestr);\n",
    "\t}\n",
    "printf(\"opening %s\\n\", fullfile);\n",
    "   \n",
    "  file = fopen( fullfile, rwstr[mode] );\n",
    "  free(fullfile);\n",
    "  return file;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "unsigned input_sequence( FILE *file, int nucleotides, int showfile )\n",
    "{\n",
    "unsigned length, i, j;\n",
    "char     c;\n",
    "#ifdef USE_MMAP\n",
    "FILE *OUTPUT;\n",
    "#endif\n",
    "\n",
    "printf(\"inputting sequence\\n\");\n",
    "\n",
    "length = 0;                           /* count how many bases */\n",
    "while( j=0, fgets( tempstr, 128, file ) != NULL )\n",
    "\t{\n",
    "    \twhile( (c=tempstr[j++]) != 0 )\n",
    "\t\t{\n",
    "      \t\tif( c == 'a'  ||  c == 't'  ||  c == 'g'  ||  c == 'c'||  c == 'n'  || c == 'A'  ||  c == 'T'  ||  c == 'G'  ||  c == 'C'||  c == 'N'  )\n",
    "\t\t\t{\n",
    "        \t\tlength ++;\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "#ifndef USE_MMAP\n",
    "sequence = (char *) malloc( length+nucleotides );\n",
    "#else\n",
    "sequence = (char *) malloc( nucleotides );\n",
    "OUTPUT = (FILE *) fopen (\"/usr/local/apache/htdocs/zhunt/temp\", \"w\");\n",
    "#endif\n",
    "\n",
    "rewind( file );\n",
    "\n",
    "if( showfile )\n",
    "\t{\n",
    "    \tprintf( \"\\n\" );\n",
    "\t}\n",
    "i = 0;\n",
    "while( j=0, fgets( tempstr, 128, file ) != NULL )\n",
    "\t{\n",
    "    \twhile( (c=tempstr[j++]) != 0 )\n",
    "\t\t{\n",
    "      \t\tif( c == 'a'  ||  c == 't'  ||  c == 'g'  ||  c == 'c' || c == 'n' ||  c == 'A' ||  c == 'T'  ||  c == 'G' ||  c == 'C' ||  c == 'N')\n",
    "        \t\t{\n",
    "\t\t\t\tif( c == 'N')\n",
    "\t\t\t\t{\n",
    "\t\t\t\tc = 'n';\n",
    "\t\t\t\t}\t\n",
    "          \t\tif( c == 'A')\n",
    "\t\t\t\t{\n",
    "\t\t\t\tc = 'a';\n",
    "\t\t\t\t}\n",
    "          \t\tif( c == 'T')\n",
    "\t\t\t\t{\n",
    "\t\t\t\tc = 't';\n",
    "\t\t\t\t}\n",
    "          \t\tif( c == 'G')\n",
    "\t\t\t\t{\n",
    "\t\t\t\tc = 'g';\n",
    "\t\t\t\t};\n",
    "          \t\tif( c == 'C')\n",
    "\t\t\t\t{\n",
    "\t\t\t\tc = 'c';\n",
    "\t\t\t\t}\n",
    "#ifndef USE_MMAP\n",
    "          \t\tsequence[i++] = c;\n",
    "#else\n",
    "\t\t\tfprintf(OUTPUT,\"%c\",c);\n",
    "\t\t\tif(i < nucleotides)\n",
    "\t\t\t\t{\n",
    "          \t\t\tsequence[i++] = c;\n",
    "\t\t\t\t}\n",
    "#endif\n",
    "        \t\t}\n",
    "\t\t}\n",
    "\t}\n",
    " for( j=0; j<nucleotides; j++ )        /* assume circular nucleotides */\n",
    "\t{\n",
    "#ifndef USE_MMAP\n",
    "    \tsequence[i++] = sequence[j];\n",
    "#else\n",
    "\tfprintf(OUTPUT,\"%c\",sequence[j]);\n",
    "#endif\n",
    "\t}\n",
    "#ifdef USE_MMAP\n",
    "\tclose(OUTPUT);\t\n",
    "\tsequencefile = open (\"/usr/local/apache/htdocs/zhunt/temp\", O_RDWR, NULL);\n",
    "\tfree(sequence);\n",
    "\tsequence = (char *) mmap(0,length, PROT_READ | PROT_WRITE,MAP_SHARED,sequencefile,0);\n",
    "#endif\n",
    "\n",
    "return length;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "/* calculate the probability of the value 'dl' in a Gaussian distribution */\n",
    "/* from \"Data Reduction and Error Analysis for the Physical Science\" */\n",
    "/* Philip R. Bevington, 1969, McGraw-Hill, Inc */\n",
    "\n",
    "\n",
    "double assign_probability( double dl )\n",
    "{\n",
    "  static double average = 29.6537135;\n",
    "  static double stdv = 2.71997;\n",
    "  static double _sqrt2 = 0.70710678118654752440;   /* 1/sqrt(2) */\n",
    "  static double _sqrtpi = 0.564189583546;  /* 1/sqrt(pi) */\n",
    "\n",
    "  double x, y, z, k, sum;\n",
    "\n",
    "  z = fabs( dl - average ) / stdv;\n",
    "  x = z * _sqrt2;\n",
    "  y = _sqrtpi * exp( -x * x );\n",
    "  z *= z;\n",
    "  k = 1.0;\n",
    "  sum = 0.0;\n",
    "  do\n",
    "    {\n",
    "      sum += x;\n",
    "      k += 2.0;\n",
    "      x *= z / k;\n",
    "    }  while( sum + x > sum );\n",
    "  z = 0.5 - y * sum;                    /* probability of each tail */\n",
    "  return (dl > average) ? z : 1.0/z;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "double find_delta_linking( int dinucleotides );\n",
    "void   calculate_zscore( double a, int maxdinucleotides, int min, int max, char *filename );\n",
    "\n",
    "void   print_array( int points[], unsigned from, unsigned to, int printwidth );\n",
    "void   soft_copy( int width, int widthbyte, char **screen );\n",
    "\n",
    "void   generate_random_sequence( int dinucleotides, char seq[] );\n",
    "void   initiate_random( long seed );\n",
    "double uniform_random( void );\n",
    "void   random_distribution( double a, int maxdinucleotides, char *filename );\n",
    "void   run_distribution( double a, int maxdinucleotides, FILE *disfile );\n",
    "void   print_distribution( char *disname, FILE *disfile );\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "double find_delta_linking( int dinucleotides )\n",
    "{\n",
    "  double sum;\n",
    "  int    i, j;\n",
    "\n",
    "  for( i=0; i<dinucleotides; i++ )\n",
    "    bzenergy[i] = 1.0;\n",
    "  for( i=0; i<dinucleotides; i++ )\n",
    "    {\n",
    "      sum = 0.0;\n",
    "      for( j=0; j<dinucleotides-i; j++ )\n",
    "        {\n",
    "          bzenergy[j] *= best_bzenergy[i+j];\n",
    "          sum += bzenergy[j];\n",
    "        }\n",
    "      logcoef[i] = log( sum );\n",
    "    }\n",
    "  terms = dinucleotides;\n",
    "  return linear_search( 10.0, 50.0, 0.001, delta_linking );\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "void calculate_zscore( double a, int maxdinucleotides, int min, int max, char *filename )\n",
    "{\n",
    "  static double pideg=57.29577951;      /* 180/pi */\n",
    "  char     *bestantisyn;\n",
    "  FILE     *file;\n",
    "  unsigned seqlength, i, j, k;\n",
    "  int      fromdin, todin, din, nucleotides;\n",
    "  long     begintime, endtime;\n",
    "  double   dl, slope, probability, bestdl;\n",
    "  float    initesum;\n",
    "\n",
    "  fromdin = min;\n",
    "  todin = max;\n",
    "  printf(\"calculating zscore\\n\");\n",
    "\n",
    "  file = open_file( 1, filename, \"\" );\n",
    "  if( file == NULL )\n",
    "\t{\n",
    "\tprintf(\"couldn't open %s!\\n\", filename);\n",
    "        return;\n",
    "\t}\n",
    "  seqlength = input_sequence( file, 2*maxdinucleotides, 0 );\n",
    "  fclose( file );\n",
    "\n",
    "  file = open_file( 0, filename, \"Z-SCORE\" );\n",
    "  if( file == NULL )\n",
    "    \t{\n",
    "#ifndef USE_MMAP\n",
    "      \tfree( sequence );\n",
    "#else\n",
    "\tmunmap(sequence, seqlength);\n",
    "\tclose(sequencefile);\n",
    "#endif\n",
    "      \treturn;\n",
    "    \t}\n",
    "\n",
    "  if( todin > maxdinucleotides )\n",
    "\t{\n",
    "    \ttodin = maxdinucleotides;\n",
    "\t}\n",
    "  if( fromdin > todin )\n",
    "\t{\n",
    "    \tfromdin = todin;\n",
    "\t}\n",
    "  nucleotides = 2 * todin;\n",
    "\n",
    "#ifndef PROB_ONLY\n",
    "  fprintf( file, \"%s %u %d %d\\n\", filename, seqlength, fromdin, todin );\n",
    "#endif\n",
    "\n",
    "  a /= 2.0;\n",
    "  initesum = 10.0 * todin;\n",
    "  bestantisyn = (char *) malloc( nucleotides+1 );\n",
    "\n",
    "  time( &begintime );\n",
    "  for( i=0; i < seqlength; i++ )\n",
    "    {\n",
    "      assign_bzenergy_index( nucleotides, sequence+i );\n",
    "      bestdl = 50.0;\n",
    "      for( din=fromdin; din<=todin; din++ )\n",
    "        {\n",
    "          best_esum = initesum;\n",
    "          deltatwist = a * (double)din;\n",
    "          antisyn[2*din] = 0;\n",
    "          anti_syn_energy( 0, din, 0.0 );           /* esum = 0.0 */\n",
    "          dl = find_delta_linking( din );  \n",
    "          if( dl < bestdl )\n",
    "            {\n",
    "              bestdl = dl;\n",
    "              strcpy( bestantisyn, best_antisyn );\n",
    "            }\n",
    "        }\n",
    "\t\tj = strlen(bestantisyn);\n",
    "#ifndef PROB_ONLY\n",
    "      slope = atan( delta_linking_slope( bestdl ) ) * pideg;\n",
    "#endif\n",
    "      probability = assign_probability( bestdl );\n",
    "#ifndef PROB_ONLY\n",
    "      fprintf( file, \"%i %i %i %7.3lf %7.3lf %le \", i+1, i+1+j, j, bestdl, slope, probability ); \n",
    "#else\n",
    "      fprintf( file, \" %7.3lf %le\\n\", bestdl, probability ); \n",
    "#endif\n",
    "\tfor( k=0; k<j; k++ )\n",
    "        fprintf( file,\"%c\", sequence[i+k] );\n",
    "        fprintf( file, \"   %s\\n\", bestantisyn );\n",
    "        \n",
    "\t}\n",
    "  time( &endtime );\n",
    "\n",
    "  free( bestantisyn );\n",
    "  fclose( file );\n",
    "  printf( \"\\n run time=%ld sec\\n\", endtime-begintime );\n",
    "#ifndef USE_MMAP\n",
    "  free( sequence );\n",
    "#else\n",
    "  munmap(sequence,seqlength);\n",
    "  close(sequencefile);\n",
    "#endif\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "void analyze_zscore(  char *filename, int low, int high)\n",
    "{\n",
    "  float    *dl, *slope, *probability;\n",
    "  unsigned seqlength, i, select;\n",
    "  char     **antisyn;\n",
    "  FILE     *file;\n",
    "  int      fromdin, todin, nucleotides;\n",
    "  \n",
    "\n",
    "  printf(\"use min/max %d %d\\n\",low,high);\n",
    "  printf(\"analyzing_zscore\\n\");\n",
    "\n",
    "  file = open_file( 1, filename, \"Z-SCORE\" );\n",
    "  if( file == NULL )\n",
    "\t{\n",
    "\tprintf(\"couldn't open %s.Z-SCORE!\\n\",filename);\n",
    "        return;\n",
    "\t}\n",
    "  fscanf( file, \"%s %u %d %d\", tempstr, &seqlength, &fromdin, &todin );\n",
    "  nucleotides = 2 * todin;\n",
    "  dl = (float *) calloc( seqlength, sizeof(float) );\n",
    "  slope = (float *) calloc( seqlength, sizeof(float) );\n",
    "  probability = (float *) calloc( seqlength, sizeof(float) );\n",
    "  antisyn = (char **) calloc( seqlength, sizeof(char *) );\n",
    "\n",
    "  for( i=0; i<seqlength; i++ )\n",
    "    {\n",
    "      fscanf( file, \"%f %f %f %s\", dl+i, slope+i, probability+i, tempstr );\n",
    "      antisyn[i] = strdup( tempstr );\n",
    "    }\n",
    "  fclose( file );\n",
    "\n",
    "  file = open_file(1, filename , \"\");\n",
    "  input_sequence( file, nucleotides, 0 );\n",
    "  fclose( file );\n",
    "/* uncomment to print probability file\n",
    " show_probability( seqlength, dl, slope, probability, sequence, antisyn,filename,low,high); \n",
    " */\n",
    "  free( dl );\n",
    "  free( slope );\n",
    "  free( probability );\n",
    "  for( i=0; i<seqlength; i++ )\n",
    "    free( antisyn[i] );\n",
    "  free( antisyn );\n",
    " \n",
    "#ifndef USE_MMAP\n",
    "  free( sequence );\n",
    "#else\n",
    "  munmap(sequence,seqlength);\n",
    "  close(sequencefile);\n",
    "#endif\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "long RAm=714025, RAa=1366, RAc=150889, RAseed=0, RAy, RAr[97];\n",
    "\n",
    "\n",
    "\n",
    "void initiate_random( long seed )\n",
    "{\n",
    "  int i;\n",
    "\n",
    "  seed = (RAc - seed) % RAm;\n",
    "  if( seed < 0 )\n",
    "    seed = -seed;\n",
    "  for( i=0; i<97; i++ )\n",
    "    RAr[i] = seed = (RAa * seed + RAc) % RAm;\n",
    "  seed = (RAa * seed + RAc) % RAm;\n",
    "  RAseed = RAy = seed;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "double uniform_random( void )\n",
    "{\n",
    "  int    i;\n",
    "\n",
    "  i = (double)97.0 * RAy / (double)RAm;\n",
    "  RAy = RAr[i];\n",
    "  RAr[i] = RAseed = (RAa * RAseed + RAc) % RAm;\n",
    "  return (double)RAy / RAm;\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "void generate_random_sequence( int dinucleotides, char seq[] )\n",
    "{\n",
    "  static char rs1[] = \"aaaattttggggcccc\";\n",
    "  static char rs2[] = \"atgcatgcatgcatgc\";\n",
    "  int i, j, r;\n",
    "\n",
    "  for( i=j=0; i<dinucleotides; i++ )\n",
    "    {\n",
    "      r = uniform_random() * 17.0;\n",
    "      seq[j++] = rs1[r];\n",
    "      seq[j++] = rs2[r];\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "void random_distribution( double a, int maxdinucleotides, char *filename )\n",
    "{\n",
    "  int  select;\n",
    "  FILE *file;\n",
    "\n",
    "  do\n",
    "    {\n",
    "      printf( \"\\n 0. Stop\" );\n",
    "      printf( \"\\n 1. Run distribution\" );\n",
    "      printf( \"\\n\\n Select: \" );\n",
    "      scanf( \"%d\", &select );\n",
    "      if( select > 0 )\n",
    "        {\n",
    "          file = open_file( select-1, filename, \"DISTRIBUTION\" );\n",
    "          if( file == NULL )\n",
    "            continue;\n",
    "          if( select == 1 )\n",
    "            run_distribution( a, maxdinucleotides, file );\n",
    "         fclose( file );\n",
    "        }\n",
    "    }  while( select != 0 );\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "void run_distribution( double a, int maxdinucleotides, FILE *disfile )\n",
    "{\n",
    "  int    i, nucleotides, dinucleotides, levels;\n",
    "  long   begintime, endtime, repeat, k, *distribution;\n",
    "  float  initesum;\n",
    "  double dl, sumdl, sumdldl, dlaverage, dlstdv, dlfrom, dlto, dlstep;\n",
    "\n",
    "  printf( \"\\n Window size (dinucleotides) and sample size: \" );\n",
    "  if( user_regret() )\n",
    "    return;\n",
    "  sscanf( tempstr, \"%d %ld\", &dinucleotides, &repeat );\n",
    "  if( dinucleotides > maxdinucleotides )\n",
    "    dinucleotides = maxdinucleotides;\n",
    "  nucleotides = 2 * dinucleotides;\n",
    "\n",
    "  printf( \"\\n FROM, TO & STEP linking differences for distribution: \" );\n",
    "  if( user_regret() )\n",
    "    return;\n",
    "  sscanf( tempstr, \"%lf %lf %lf\", &dlfrom, &dlto, &dlstep );\n",
    "\n",
    "  sequence = (char *) malloc( nucleotides );\n",
    "                                   /* half of the window convert from B to Z */\n",
    "  deltatwist = a * (double)dinucleotides / 2.0;\n",
    "  initesum = 10.0 * dinucleotides;      /* for best_anti_syn() use */\n",
    "\n",
    "  levels = (dlto - dlfrom) / dlstep + 1;\n",
    "  distribution = (long *) calloc( levels, sizeof(long) );\n",
    "  for( i=0; i<levels; i++ )\n",
    "    distribution[i] = 0;\n",
    "  sumdl = sumdldl = 0.0;\n",
    "\n",
    "  time( &begintime );\n",
    "  initiate_random( begintime );\n",
    "\n",
    "  for( k=0; k<repeat; k++ )\n",
    "    {\n",
    "      generate_random_sequence( dinucleotides, sequence );\n",
    "      assign_bzenergy_index( nucleotides, sequence );\n",
    "      best_esum = initesum;\n",
    "      anti_syn_energy( 0, dinucleotides, 0.0 );\n",
    "      dl = find_delta_linking( dinucleotides );\n",
    "      sumdl += dl;\n",
    "      sumdldl += dl * dl;\n",
    "      i = (dl - dlfrom) / dlstep;\n",
    "      if( i < 0 )\n",
    "        i = 0;\n",
    "      if( i > levels - 1 )\n",
    "        i = levels - 1;\n",
    "      distribution[i] ++;\n",
    "    }\n",
    "  time( &endtime );\n",
    "  dlaverage = sumdl / repeat;\n",
    "  dlstdv = sqrt( (sumdldl - sumdl * dlaverage) / repeat );\n",
    "\n",
    "  fprintf( disfile, \" %d %d %ld %lf %lf %lg %lg %lg\\n\",\n",
    "      dinucleotides, levels, repeat, dlaverage, dlstdv, dlfrom, dlto, dlstep );\n",
    "  for( i=0; i<levels; i++ )\n",
    "    fprintf( disfile, \"%ld\\n\", distribution[i] );\n",
    "\n",
    "  printf( \"\\n run time: %ld sec\\n\", endtime-begintime );\n",
    "  free( distribution );\n",
    "  free( sequence );\n",
    "}\n",
    "\n",
    "\n",
    "void show_probability( unsigned seqlength,float *dl, float *slope, float *probability, char *sequence, char **antisyn, char *filename,int low, int high)\n",
    "{\n",
    "  unsigned i, j, k;\n",
    "  FILE     *file;\n",
    "  \n",
    "  printf(\"show_probability\\n\");\n",
    "  printf(\"min/max %d %d\\n\",low,high);\n",
    "\n",
    "  file = open_file( 0, filename, \"probability\" );\n",
    " \n",
    "  if( file == NULL )\n",
    "\t{\n",
    "\tprintf(\"couldn't open %s.probability!\\n\",filename);\n",
    "        return;\n",
    "\t}\n",
    "\tfprintf( file, \"%s %u %i %i\\n\", filename, seqlength, low, high );\n",
    "  for( i=0; i < seqlength; i++ )\n",
    "        {\n",
    "\t\tj = strlen( antisyn[i] );\n",
    "\t\tfprintf( file, \" %i %i %i %7.3f %7.3f  %10.3e  \", i+1, i+1+j, j, dl[i], slope[i], probability[i] );\n",
    "        \n",
    "        for( k=0; k<j; k++ )\n",
    "        fprintf( file, \"%c\", sequence[i+k] );\n",
    "        fprintf( file, \"   %s\\n\", antisyn[i] );\n",
    "        }\n",
    "\n",
    " fclose( file );\n",
    "}\n",
    "\n",
    "\n",
    "int main( int argc, char *argv[])\n",
    "{\n",
    "    static double rt=0.59004;             /* 0.00198*298 */\n",
    "    static double a=0.357, b=0.4;         /* a = 2 * (1/10.5 + 1/12) */\n",
    "    double   ab;\n",
    "    int      i, j, nucleotides, dinucleotides, select;\n",
    "    int min, max;\n",
    "    int low, high;\n",
    "\n",
    "    if(argc < 5)\n",
    "    {\n",
    "        printf(\"usage: zhunt windowsize minsize maxsize datafile\\n\");\n",
    "        exit(1);\n",
    "    }\n",
    "    tempstr = (char *) malloc( 128 );\n",
    "    dinucleotides = atoi((char *)argv[1]);\n",
    "    min = atoi((char *)argv[2]);\n",
    "    max = atoi((char *)argv[3]);\n",
    "    low = atoi((char *)argv[2]);\n",
    "    high = atoi((char *)argv[3]);\n",
    "\n",
    "    printf(\"dinucleotides %d\\n\", dinucleotides);\n",
    "    printf(\"min/max %d %d\\n\", min, max);\n",
    "    printf(\"min/max %d %d\\n\", low, high);\n",
    "    printf(\"operating on %s\\n\", (char *)argv[4]);\n",
    "\n",
    "    nucleotides = 2 * dinucleotides;\n",
    "\n",
    "    antisyn = (char *) malloc( nucleotides+1 );\n",
    "    best_antisyn = (char *) malloc( nucleotides+1 );\n",
    "\n",
    "    bzindex = (int *) calloc( dinucleotides, sizeof(int) );\n",
    "    bzenergy = (double *) calloc( dinucleotides, sizeof(double) );\n",
    "    best_bzenergy = (double *) calloc( dinucleotides, sizeof(double) );\n",
    "\n",
    "    bztwist = (double *) calloc( dinucleotides, sizeof(double) );\n",
    "    ab = b + b;\n",
    "    for( i=0; i<dinucleotides; i++ )\n",
    "    {\n",
    "        ab += a;\n",
    "        bztwist[i] = ab;\n",
    "    }\n",
    "\n",
    "    for( i=0; i<4; i++ )\n",
    "        for( j=0; j<17; j++ )\n",
    "            expdbzed[i][j] = exp( -dbzed[i][j] / rt );\n",
    "\n",
    "    logcoef = (double *) calloc( dinucleotides, sizeof(double) );\n",
    "    exponent = (double *) calloc( dinucleotides, sizeof(double) );\n",
    "\n",
    "    calculate_zscore( a, dinucleotides, min, max, (char *)argv[4] );\n",
    "#ifndef PROB_ONLY\n",
    "    analyze_zscore((char *)argv[4],low,high);\n",
    "#endif\n",
    "\n",
    "    free( exponent );\n",
    "    free( logcoef );\n",
    "    free( bztwist );\n",
    "    free( best_bzenergy );\n",
    "    free( bzenergy );\n",
    "    free( bzindex );\n",
    "    free( best_antisyn );\n",
    "    free( antisyn );\n",
    "    free( tempstr );\n",
    "    return 0;\n",
    "}\n",
    "\n",
    "\"\"\"\n",
    "with open(\"zhunt3-alan.c\", \"w\", encoding=\"utf-8\") as f:\n",
    "    f.write(zhunt_impl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-17T12:58:59.114412Z",
     "iopub.status.busy": "2025-06-17T12:58:59.114080Z",
     "iopub.status.idle": "2025-06-17T12:58:59.696416Z",
     "shell.execute_reply": "2025-06-17T12:58:59.694980Z",
     "shell.execute_reply.started": "2025-06-17T12:58:59.114390Z"
    },
    "trusted": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clang: \u001b[0;1;35mwarning: \u001b[0m\u001b[1margument '-Ofast' is deprecated; use '-O3 -ffast-math' for the same behavior, or '-O3' to enable only conforming optimizations [-Wdeprecated-ofast]\u001b[0m\n",
      "\u001b[1mzhunt3-alan.c:52:11: \u001b[0m\u001b[0;1;35mwarning: \u001b[0m\u001b[1mpassing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C23 [-Wdeprecated-non-prototype]\u001b[0m\n",
      "   52 |   f = func( x1 );\u001b[0m\n",
      "      | \u001b[0;1;32m          ^\n",
      "\u001b[0m\u001b[1mzhunt3-alan.c:53:14: \u001b[0m\u001b[0;1;35mwarning: \u001b[0m\u001b[1mpassing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C23 [-Wdeprecated-non-prototype]\u001b[0m\n",
      "   53 |   fmid = func( x2 );\u001b[0m\n",
      "      | \u001b[0;1;32m             ^\n",
      "\u001b[0m\u001b[1mzhunt3-alan.c:61:18: \u001b[0m\u001b[0;1;35mwarning: \u001b[0m\u001b[1mpassing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C23 [-Wdeprecated-non-prototype]\u001b[0m\n",
      "   61 |       fmid = func( xmid );\u001b[0m\n",
      "      | \u001b[0;1;32m                 ^\n",
      "\u001b[0m\u001b[1mzhunt3-alan.c:304:7: \u001b[0m\u001b[0;1;35mwarning: \u001b[0m\u001b[1m'gets' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of gets(3), it is highly recommended that you use fgets(3) instead. [-Wdeprecated-declarations]\u001b[0m\n",
      "  304 |       gets( tempstr );\u001b[0m\n",
      "      | \u001b[0;1;32m      ^\n",
      "\u001b[0m\u001b[1m/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_stdio.h:260:1: \u001b[0m\u001b[0;1;36mnote: \u001b[0m'gets' has been explicitly marked deprecated here\u001b[0m\n",
      "  260 | __deprecated_msg(\u001b[0;32m\"This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of gets(3), it is highly recommended that you use fgets(3) instead.\"\u001b[0m)\u001b[0m\n",
      "      | \u001b[0;1;32m^\n",
      "\u001b[0m\u001b[1m/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:218:48: \u001b[0m\u001b[0;1;36mnote: \u001b[0mexpanded from macro '__deprecated_msg'\u001b[0m\n",
      "  218 |         #define __deprecated_msg(_msg) \u001b[0;34m__attribute__\u001b[0m((__deprecated__(_msg)))\u001b[0m\n",
      "      | \u001b[0;1;32m                                                      ^\n",
      "\u001b[0m4 warnings generated.\n"
     ]
    }
   ],
   "source": [
    "!gcc zhunt3-alan.c -Ofast -lm -o zhunt3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-17T12:59:32.391316Z",
     "iopub.status.busy": "2025-06-17T12:59:32.390873Z"
    },
    "id": "34Rsg-V5obZV",
    "trusted": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-06-18 00:23:44--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/022/581/195/GCF_022581195.2_ilHelZeax1.1/GCF_022581195.2_ilHelZeax1.1_genomic.fna.gz\n",
      "Распознаётся ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)… 130.14.250.10, 130.14.250.31, 130.14.250.13, ...\n",
      "Подключение к ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... соединение установлено.\n",
      "HTTP-запрос отправлен. Ожидание ответа… 200 OK\n",
      "Длина: 120848234 (115M) [application/x-gzip]\n",
      "Сохранение в: «GCF_022581195.2_ilHelZeax1.1_genomic.fna.gz»\n",
      "\n",
      "GCF_022581195.2_ilH 100%[===================>] 115,25M  2,07MB/s    за 45s     \n",
      "\n",
      "2025-06-18 00:24:29 (2,59 MB/s) - «GCF_022581195.2_ilHelZeax1.1_genomic.fna.gz» сохранён [120848234/120848234]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/022/581/195/GCF_022581195.2_ilHelZeax1.1/GCF_022581195.2_ilHelZeax1.1_genomic.fna.gz\n",
    "!gunzip ./GCF_022581195.2_ilHelZeax1.1_genomic.fna.gz\n",
    "!mv GCF_022581195.2_ilHelZeax1.1_genomic.fna genomic.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "id": "gEUQ3dq3omL0"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dinucleotides 12\n",
      "min/max 8 12\n",
      "min/max 8 12\n",
      "operating on genomic.fna\n",
      "calculating zscore\n",
      "opening genomic.fna\n",
      "inputting sequence\n",
      "opening genomic.fna.Z-SCORE\n",
      "\n",
      " run time=13342 sec\n",
      "use min/max 8 12\n",
      "analyzing_zscore\n",
      "opening genomic.fna.Z-SCORE\n",
      "opening genomic.fna\n",
      "inputting sequence\n"
     ]
    }
   ],
   "source": [
    "!./zhunt3 12 8 12 genomic.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "yb0tycl8orUG"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genomic.fna 375181922 8 12\n",
      "1 17 16  32.498  14.812 1.478861e-01 nccaaattacaccagn   ASASASSASASAASAS\n",
      "2 18 16  30.515  17.723 3.757344e-01 ccaaattacaccagnt   ASASASASASASASAS\n",
      "3 19 16  31.913  25.037 2.031080e-01 caaattacaccagntg   ASASSASAASASASAS\n",
      "4 20 16  30.501  17.603 3.776943e-01 aaattacaccagntgn   ASASASASASASASAS\n",
      "5 23 18  31.545  28.752 2.434462e-01 aattacaccagntgnncc   SASASAASASASASASAS\n",
      "6 22 16  30.500  17.222 3.777796e-01 attacaccagntgnnc   ASASASASASASASAS\n",
      "7 23 16  31.187  31.021 2.864575e-01 ttacaccagntgnncc   SASAASASASASASAS\n",
      "8 24 16  30.507  17.142 3.768418e-01 tacaccagntgnnccc   ASASASASASASASAS\n",
      "9 25 16  31.006  28.613 3.094738e-01 acaccagntgnnccct   SAASASASASSASASA\n"
     ]
    }
   ],
   "source": [
    "!head genomic.fna.Z-SCORE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "id": "O5qOQPc5otBC"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_061452.1': (1, 15512169),\n",
       " 'NC_061453.1': (15512170, 30573753),\n",
       " 'NC_061454.1': (30573754, 45268912),\n",
       " 'NC_061455.1': (45268913, 59434628),\n",
       " 'NC_061456.1': (59434629, 73485891),\n",
       " 'NC_061457.1': (73485892, 87509142),\n",
       " 'NC_061458.1': (87509143, 101159065),\n",
       " 'NC_061459.1': (101159066, 114808336),\n",
       " 'NC_061460.1': (114808337, 128423600),\n",
       " 'NC_061461.1': (128423601, 141681985),\n",
       " 'NC_061462.1': (141681986, 154770825),\n",
       " 'NC_061463.1': (154770826, 167743605),\n",
       " 'NC_061464.1': (167743606, 180584909),\n",
       " 'NC_061465.1': (180584910, 193342856),\n",
       " 'NC_061466.1': (193342857, 205948777),\n",
       " 'NC_061467.1': (205948778, 218281433),\n",
       " 'NC_061468.1': (218281434, 230536315),\n",
       " 'NC_061469.1': (230536316, 242781788),\n",
       " 'NC_061470.1': (242781789, 254411634),\n",
       " 'NC_061471.1': (254411635, 266011388),\n",
       " 'NC_061472.1': (266011389, 277472905),\n",
       " 'NC_061473.1': (277472906, 288343946),\n",
       " 'NC_061474.1': (288343947, 298766083),\n",
       " 'NC_061475.1': (298766084, 308508591),\n",
       " 'NC_061476.1': (308508592, 317789078),\n",
       " 'NC_061477.1': (317789079, 327067491),\n",
       " 'NC_061478.1': (327067492, 334422524),\n",
       " 'NC_061479.1': (334422525, 341634324),\n",
       " 'NC_061480.1': (341634325, 348837802),\n",
       " 'NC_061481.1': (348837803, 355154615),\n",
       " 'NC_061482.1': (355154616, 373959895),\n",
       " 'NW_025899711.1': (373959896, 374459896),\n",
       " 'NW_025899712.1': (374459897, 374709896),\n",
       " 'NW_025899713.1': (374709897, 374834917),\n",
       " 'NW_025899714.1': (374834918, 374923073),\n",
       " 'NW_025899715.1': (374923074, 374974850),\n",
       " 'NW_025899716.1': (374974851, 375022136),\n",
       " 'NW_025899717.1': (375022137, 375064453),\n",
       " 'NW_025899718.1': (375064454, 375099399),\n",
       " 'NW_025899719.1': (375099400, 375124416),\n",
       " 'NW_025899720.1': (375124417, 375145340),\n",
       " 'NW_025899721.1': (375145341, 375165593),\n",
       " 'NC_061507.1': (375165594, 375180944)}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seg_dict = {}\n",
    "sum = 1\n",
    "input_file = \"genomic.fna\"\n",
    "fasta_sequence = SeqIO.parse(input_file,'fasta')\n",
    "\n",
    "for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "    name, sequence = record.id, str(record.seq)\n",
    "    seg_dict[name] = (sum, sum + len(sequence) - 1)\n",
    "    sum += len(sequence)\n",
    "seg_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "id": "6Il7-S7Wouf-"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: results_zhunt.bed: No such file or directory\n",
      "NC_061452.1\t3538\t3554\n",
      "NC_061452.1\t4526\t4542\n",
      "NC_061452.1\t9982\t9998\n",
      "NC_061452.1\t10589\t10607\n",
      "NC_061452.1\t10591\t10607\n",
      "NC_061452.1\t10597\t10615\n",
      "NC_061452.1\t10598\t10616\n",
      "NC_061452.1\t14209\t14229\n",
      "NC_061452.1\t14211\t14229\n",
      "NC_061452.1\t14212\t14232\n"
     ]
    }
   ],
   "source": [
    "!rm 'results_zhunt.bed'\n",
    "\n",
    "with open('genomic.fna.Z-SCORE', 'r') as f:\n",
    "    ind = -1\n",
    "    while 1:\n",
    "        ind += 1\n",
    "        line = f.readline()\n",
    "        if not line:\n",
    "            break\n",
    "        if ind < 1:\n",
    "            continue\n",
    "        start = int(line.split()[0])\n",
    "        end = int(line.split()[1])\n",
    "        score = float(line.split()[5])\n",
    "        name_start = ''\n",
    "        name_end = ''\n",
    "\n",
    "        if score < 400:\n",
    "            continue\n",
    "\n",
    "        for key, (l, r) in seg_dict.items():\n",
    "            if start >= l and start <= r:\n",
    "                name_start = key\n",
    "                start -= l\n",
    "                if end >= l and end <= r:\n",
    "                    name_end = key\n",
    "                    end -= l\n",
    "                break\n",
    "\n",
    "        if name_start != name_end:\n",
    "            continue\n",
    "\n",
    "        if name_start == '':\n",
    "            continue\n",
    "\n",
    "        with open('results_zhunt.bed', 'a') as g:\n",
    "            g.write(f'{name_start}\\t{start}\\t{end}\\n')\n",
    "\n",
    "!head 'results_zhunt.bed'"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kaggle": {
   "accelerator": "none",
   "dataSources": [],
   "dockerImageVersionId": 31040,
   "isGpuEnabled": false,
   "isInternetEnabled": false,
   "language": "python",
   "sourceType": "notebook"
  },
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
